torch/csrc/jit/codegen/cuda/kernel_db/kernel_db.cpp

```diff
@@ -0,0 +1,266 @@
+#include <mutex>
+#include <regex>
+
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_db/kernel_db.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_db/utils.h>
+#include <torch/csrc/jit/codegen/cuda/utils.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+static std::mutex kernel_db_lock;
+
+KernelDb::KernelDb(bool _disabled)
+    : disabled_(_disabled),
+      initialized_(false),
+      kernel_map_(),
+      kernel_db_path_(),
+      kernel_db_txt_file_() {}
+
+KernelDb& KernelDb::get() {
+  const std::string kernel_db_dir = "nvfuser_kernel_db";
+  const std::string kernel_db_file = "db.csv";
+
+  return get(
+      kernel_db_dir,
+      kernel_db_file,
+      true,
+      !isOptionEnabled(EnableOption::KernelDb),
+      false);
+}
+
+KernelDb& KernelDb::get(
+    const std::string& kernel_db_dir,
+    const std::string& kernel_db_file,
+    bool use_temp_dir,
+    bool disabled,
+    bool reset) {
+  std::lock_guard<std::mutex> guard(kernel_db_lock);
```

Comment:
I have concern over the thread-safe claim on KernelDb.

We are only locking mutex inside `KernelDb::get` and `KernelDb::write`. But `KernelDb::get` can return the instance, which would blow our thread safety here, since it's now readable and there's nothing preventing that to overlap with the write.